Add a guide to deploying an HA cluster

ptgott opened the pull request:
Closes #16751

This is a general guide that prefaces our HA deployment guides.

Also adds an introduction to the "deploy-a-cluster" section.

## Notes for reviewers:

- Once the text of the guide looks okay, I'll add a diagram within this PR. The complete PR will include the text and the diagram.
- @hugoShaka Is editing the `teleport-cluster` chart to deploy the Auth Service and Proxy Service as separate groups. Since this is still being implemented, I thought we would start by describing the "old" best practice then, when the new `teleport-cluster` chart is available, either replacing this approach with that one or somehow incorporating both approaches.
- I have a separate PR (https://github.com/gravitational/teleport/pull/19205) to make the "Deploy a Cluster" section only for reference deployment guides, moving our guides to Teleport's editions (Cloud, Enterprise) into a separate section. The HA guide will go in the "Deploy a Cluster" section.

ptgott committed:
Add a guide to deploying an HA cluster

Closes #16751

This is a general guide that prefaces our HA deployment guides.

Also adds an introduction to the "deploy-a-cluster" section.

Valien commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+### Deploy Teleport services
+
+Once your high-availability Teleport deployment is up and running, you can add
+2esources by launching Teleport services. You can run these services in a
```
Typo here. Should be **resources**

hugoShaka commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+
+| Port | Description |
+| - | - |
+| `3021` | Port used by Teleport Proxy Service instances to dial agents in [Proxy Peering mode](../management/operations/proxy-peering.mdx). |
```
AFAIK this port is only for inter-proxy communications and should not be exposed publicly nor load-balanced.

ptgott replied:
👍 I've deleted this line from each of the port tables.

Should Proxy Peering be enabled in the architecture described here? From my cursory understanding of Proxy Peering, it seems like, if agents are connecting to a Proxy Service instance via the load balancer, they don't need to be aware of individual Proxy Service instances, so Proxy Peering wouldn't be relevant here.

hugoShaka replied:
I don't know if we should talk or not about proxy peering here, mainly because this is an enterprise feature and is still in preview.

However, agents are aware of individual proxy instances even through a loadbalancer 🙃 During their first connection they will discover how many proxies are connected to the cluster, and will start brute-forcing their way through the LB to have an open connection with each proxy. If they end up on a proxy they are already connected to, the proxy returns a "try again" message. [At scale, this can take some time for an agent to reconnect to all proxies.](https://github.com/gravitational/teleport/issues/13129#issuecomment-1249404054)

Without proxy peering, if a user ends up on a proxy the agent is not connected to, they will receive an error. With proxy peering, the proxy can lookup if the resource is available through another proxy. If it is, it's going to forward the connection to the agent's proxy on the proxy peering port. This capability also means the agents don't have to maintain an open connection per proxy, they can just keep 2 or 3 connections open and still be reachable.

ptgott replied:
Thanks for clarifying!

webvictim replied:
I agree that proxy peering is an advanced, Enterprise-only use case that probably shouldn't be talked about in the context of this guide. It's still early days and we don't have a ton of people using it yet.

hugoShaka commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+
+### Open ports
+
+Ensure that, on each Teleport instance, the following ports allow traffic from
+the load balancer. If you do not plan to run a service associated with a given
+port, you can plan to leave that port closed.
```
There are two modes to expose Teleport:
- separate listener which is the legacy way of doing, is super confusing, but most users are using it. This is the one described in this guide.
- multiplexing: the "new" one, one port to rule them all: 443 (or 3080). Ideally we would like to steer users toward the multiplexing mode (it's not the default for Helm charts because a lot of users are using a L7 lb for the 443 port and we don't support yet to pass through, hopefully this should happen in the next months and our life will become easier)

Do we want this section to reflect both modes? If we do, we might want to put multiplexing forward as it's easier for everyone.

pschisa replied:
On the implementation team side, we generally recommend users opt for TLS routing unless there is an overriding reason to not use TLS routing as it simplifies the configuration of the infra and minimizes the port exposure of Teleport

webvictim replied:
I personally don't really ever recommend TLS routing because >75% of the people I talk to want to put a load balancer of some kind in front of Teleport. >50% of the people want to use ACM. Lots of them want to use a reverse proxy like nginx or Traefik.

Probably around 10-15% want to use their own certs, and the rest are OK with Let's Encrypt. It's very popular, but just not a viable option in many environments.

This will definitely change if and when it becomes easier to keep ALPN/SNI intact behind load balancers.

ptgott replied:
Since the question of whether to enable TLS Routing is implementation specific, I'll describe both approaches in this guide!

hugoShaka commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+| Port | Description |
+| - | - |
+| `3021` | Port used by Teleport Proxy Service instances to dial agents in [Proxy Peering mode](../management/operations/proxy-peering.mdx). |
+| `3022` | SSH port. This is Teleport's equivalent of port `22` for SSH. |
```
AFAIK this port should not be exposed on the Teleport cluster, it is used for the client or the proxy to connect to a node when there's no reverse tunnel available.

hugoShaka commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+| `3022` | SSH port. This is Teleport's equivalent of port `22` for SSH. |
+| `3023` | SSH port for clients connect to. The Proxy Service will forward this connection to port `3022` on the destination Node. |
+| `3024` | SSH port used to create reverse SSH tunnels from behind-firewall environments. |
+| `3025` | TLS port used by the Auth Service to serve its gRPC API to Teleport services  in a cluster. |
```
In most setups this port should not be exposed. If the proxy is exposed and has connectivity to auth nodes, clients should never contact the auth directly.

There are a few edge cases where this can be needed, mostly because of some plugins not supporting connection through proxy. For setups with proxies it is not advised to expose auth directly.

pschisa replied:
There is also the use case for scalability, as a single cluster that exposes the auth service to the resources for node joining can support 5x the amount of nodes that our proxy service joining can.  Larger implementations have opted for joining nodes to the auth service.

I agree though that the default should be joining through the proxy with private auth servers.  Of course, our current helm deployment has both the proxy and auth service running in the same pod so that muddies the waters here a bit too

ptgott replied:
Makes sense! I'll remove this line from the table. I imagine that customers with specific use cases will be okay adopting a different architecture than this guide describes in conversation with the Teleport team. If users keep running into issues with scaling because of this guide, we can think about a way to include instructions for connecting to the Auth Service directly.

hugoShaka commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+| `3025` | TLS port used by the Auth Service to serve its gRPC API to Teleport services  in a cluster. |
+| `443` | HTTPS connections to authenticate `tsh` users into the cluster. The same connection is used to serve a Web UI. |
+| `3026` | HTTPS Kubernetes proxy |
+| `3027` | Kubernetes Service |
```
Like 3022, this port should not be exposed directly by a Telelport cluster, this is used for client and proxies to contact a kubernetes service when no reverse tunnel is available.

Plus it's extra confusing as we default to 3026 but in certain cases we need to run both the kubernetes proxy service and the kubernetes service on the same machine. In this case the kubernetes proxy service keeps the 3026 and the kubernetes service takes the 3027. Clients wanting to connect to kubernetes would contact the proxy on 3026 and the proxy would dial localhost:3027.

hugoShaka commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+| `443` | HTTPS connections to authenticate `tsh` users into the cluster. The same connection is used to serve a Web UI. |
+| `3026` | HTTPS Kubernetes proxy |
+| `3027` | Kubernetes Service |
+| `3028` | Desktop Service |
```
I'm not sure if this is the same case as 3022 and 3027. It sounds like.

webvictim replied:
Yes, I believe so; `windows_desktop_service` was pretty much initially a copy and paste with `s/kubernetes_service/windows_desktop_service/g` and `s/3027/3028/g`

hugoShaka commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+
+This configuration has the following required settings:
+
+- **Disable TLS multiplexing:** This configuration uses the `v1` version of the
+  Teleport configuration (`version: v1`) to disable [TLS
+  multiplexing](../management/operations/tls-routing.mdx), which is required
+  when running Teleport with a load balancer in high-availability mode. 
```
I don't think this is required nor a good practice. v1 keeps multiplexing disabled for compatibility reasons, and the Helm charts defaults to separate listeners to stay backward compatible for users running on EKS with ACM and a LB terminating TLS on 443.

webvictim replied:
IIUC using `version: v1` in the config does two things:

- disables TLS multiplexing, equivalent to doing this on `v2`+:
```yaml
auth_service:
  proxy_listener_mode: separate
```
- automatically enables the legacy listeners on separate ports for any enabled services:
```yaml
proxy_service:
  enabled: true
  listen_addr: 0.0.0.0:3023
  tunnel_listen_addr: 0.0.0.0:3024
```

Essentially it's a shortcut to doing less legwork in the config file. It's currently the easiest way I know of to get a "legacy"/separate port config without having to be very explicit about it.

hugoShaka reviewed (commented):
Thank you for writing this, this is really needed 🙌 

I added a couple of comments regarding ports, loadbalancing and listener configuration.

pschisa commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+- **A Layer Four load balancer** to direct traffic from users and services to an
+  available `teleport` process.
+- A **key-value store** for Auth Service state and audit events that all
+  `teleport` instances can access. This requires permissions for `teleport`
```
I would call them `auth services` or auth instances` to be clear it is only the auth service process that needs to reach the key-value store

Same for the object storage service and license file sections

ptgott replied:
👍 I've mentioned the specific services throughout the guide

pschisa commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+The load balancer forwards traffic from users and services to an available
+Teleport instance. This must not terminate TLS, and must transparently forward
+the TCP traffic it receives. In other words, this must be a Layer Four load
+balancer, not a Layer Seven (e.g., HTTP) load balancer. 
```
We do support ACM and TLS termination on ALBs and other load balancers for the web port, though not with TLS multiplexing enabled

pschisa commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+For Amazon DynamoDB and Google Cloud Firestore, your Teleport configuration (which
+we will describe in more detail in the [Configuration](#configuration) section)
+names a table or collection where Teleport stores cluster state and audit
+events. For etcd, Teleport uses namespaces within item keys to identify audit
```
etcd is not supported for audit event storage, we have no on-prem solution for audit event storage at this time

pschisa commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+
+High-availability Teleport deployments use an object storage service for
+persisting session recordings. The Teleport Auth Service supports two object
+storage services:
```
We support any S3 compatible object storage.  On-prem customers often opt for MinIO or Ceph as solutions outside of the major cloud providers

pschisa commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+## TLS credential provisioning 
+
+High-availability Teleport deployments require an automated system to fetch
+certificates from Let's Encrypt and provision Teleport instances with them. 
```
Do we want to stay more generic than lets encrypt?  ACM, Digicert and self signed certificates are not uncommon alternatives to using LetsEncypt

pschisa commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,369 @@
+Set up a DNS zone where you can create records for Teleport, e.g., an Amazon
+Route 53 hosted zone or Google Cloud DNS zone.
+
+You must create a DNS A record that points to the address of your load balancer. 
```
small Nit but in AWS if assigning a load balancer as the destination IIRC it uses an Alias and that is more like a CNAME record than A record.  Stating you must create an A record could create confusion.

ptgott replied:
Thanks! What DNS records do HA setups usually include on AWS, Google Cloud, and on-prem deployments?

webvictim replied:
In pretty much every case the minimum needed is either an A record or CNAME from `teleport.example.com` to the public load balancer IP or DNS name in front of the Teleport proxies.

(AWS' Route 53 also has the concept of aliases which are basically an internally-expanded CNAME with optional health checks; for all intents and purposes the outcome is the same as a CNAME to the public load balancer DNS name)

When using application access we also encourage people to add the same for `*.teleport.example.com`

ptgott committed:
Respond to some PR feedback

- Add mentions of specific services (Auth and Proxy) where it would help
  with clarity.
- Edit port tables to remove ports that shouldn't be publicly exposed.
- Clarify the supported backends.
- Clarify that Let's Encrypt is not required for TLS credential
  management.

ptgott commented:
I've addressed the more straightforward feedback that I could implement without rethinking any structural aspects of the page. I'll work on the rest later today/this week!

webvictim commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,368 @@
+
+This includes:
+
+- **A Layer Four load balancer** to direct traffic from users and services to an
```
```suggestion
- **A Layer 4 load balancer** to direct traffic from users and services to an
```

Might just be me, but I'm not sure anyone in the history of the internet has ever talked about OSI Layer Four - it's always `Layer 4` or `L4`. See https://en.wikipedia.org/wiki/OSI_model etc. It's jarring to read the word rather than the number.

I assume this is a deliberate thing as it's done throughout this guide, but it looks very incongruous. We use `layer 4` and `layer 7` everywhere else in the docs so I'd encourage us to do the same here.

ptgott replied:
Using words for numbers under 10 has been beaten into me through years of adhering to style guides. I'll change the layer names to use numerals.

webvictim commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,368 @@
+
+| Port | Description |
+| - | - |
+| `3023` | SSH port for clients connect to. The Proxy Service will forward this connection to port `3022` on the destination Node. |
```
When you're connecting to a node joined over a reverse tunnel, port 3022 isn't used at all. The proxy routes your connection to the correct reverse tunnel and provides you with a transport channel to the node which flows over that reverse tunnel.

webvictim commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,368 @@
+| `3023` | SSH port for clients connect to. The Proxy Service will forward this connection to port `3022` on the destination Node. |
+| `3024` | SSH port used to create reverse SSH tunnels from behind-firewall environments. |
+| `443` | HTTPS connections to authenticate `tsh` users into the cluster. The same connection is used to serve a Web UI. |
+| `3026` | HTTPS Kubernetes proxy |
+| `3028` | Desktop Service |
+| `3036` | MySQL port |
+| `5432` | Postgres port |
```
We might want to note that all these ports are optional - i.e. if you're not using desktop/Kubernetes/database access you don't need them. 443, 3023 and 3024 are pretty much mandatory but the others aren't.

webvictim commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,368 @@
+Teleport demonstrates that it controls the ALPN server at the HTTPS address of
+your Teleport Proxy Service.
+
+For high-availability deployments that use Let's Encrypt to supply TLS
+credentials to Teleport instances running behind a load balancer, you will need
+to use the [ACME
+DNS-01](https://letsencrypt.org/docs/challenge-types/#dns-01-challenge)
+challenge to demonstrate domain name ownership to Let's Encrypt. In this
+challenge, your TLS credential provisioning system creates a DNS TXT record with
+a value expected by Let's Encrypt.
```
A DNS-01 challenge must also be used to get a wildcard cert from Let's Encrypt (which may be useful in the context of Application Access)

ptgott replied:
Testing a single-instance Teleport deployment with one application registered (this guide: https://goteleport.com/docs/deploy-a-cluster/open-source/), it looks like Teleport requests a certificate for each registered application's subdomain individually from Let's Encrypt, rather than directly requesting a cert for a wildcard domain. Are there instances where Teleport does request a cert for a wildcard domain?

webvictim replied:
Yes, you're absolutely right that Teleport's built-in ACME will get certs for each individual application and maintain those separately. Teleport itself never requests wildcard certs, so it's probably not relevant here!

webvictim commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,368 @@
+## Teleport instances
+
+Run the Teleport Auth Service and Proxy Service as a scalable group of compute
+resources, for example, a Kubernetes `Deployment`  or AWS Auto Scaling group.
```
```suggestion
resources, for example, a Kubernetes `Deployment` or AWS Auto Scaling group.
```

webvictim commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,368 @@
+| `3024` | SSH port used to create reverse SSH tunnels from behind-firewall environments. |
+| `443` | HTTPS connections to authenticate `tsh` users into the cluster. The same connection is used to serve a Web UI. |
+| `3026` | HTTPS Kubernetes proxy |
+| `3028` | Desktop Service |
+| `3036` | MySQL port |
+| `5432` | Postgres port |
```
Same comment as above regarding these being optional

webvictim commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,368 @@
+<Admonition title="Required permissions">
+
+If you are using Let's Encrypt to provide TLS credentials to your Teleport
+instances, the TLS credential system  we mentioned earlier needs permissions to
```
```suggestion
instances, the TLS credential system we mentioned earlier needs permissions to
```

webvictim commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,368 @@
+- **Configure storage:** The `storage` section configures the Auth Service
+  backend. The `audit_sessions_uri` names the GCS or S3 bucket to use for
+  session recordings, and the remaining fields in `storage` are specific to
+  each backend.  Consult our [Backends Reference](../reference/backends.mdx)
+  for the values you should use.  
```
```suggestion
  each backend. Consult our [Backends Reference](../reference/backends.mdx)
  for the values you should use.
```

webvictim commented:
I love this guide. It's definitely needed to bridge some gaps we have.

ptgott committed:
Respond to more PR feedback

- Add context around DNS records
- Use "Layer 4" instead of "Layer Four"
- Explain optional ports more explicitly
- Fix spacing issues
- Indicate that you can use an S3-compatible object store
- Clarify cert-fetching behavior for applications
- Describe separate port configs depending on whether TLS Routing is
  enabled, and add a brief section re: whether to enable TLS Routing
- Expand the Teleport configuration section to accommodate TLS Routing
  and separate listeners

ptgott committed:
Add more context to the "Deploy a Cluster" intro

ptgott committed:
Small language tweaks

ptgott committed:
Add an image

zmb3 commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,520 @@
+
+- **A Layer 4 load balancer** to direct traffic from users and services to an
+  available `teleport` process.
+- A **key-value store** for Auth Service state and audit events that all Auth
```
I might find a way to add the terms "cluster state" or "backend" to this item, as these terms are commonly used throughout the docs to describe the key-value store.

zmb3 commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,520 @@
+- A **key-value store** for Auth Service state and audit events that all Auth
+  Service instances can access. This requires permissions for Auth Service
+  instances to manage records within the key-value store.
+- An **object storage service** for session recordings uploaded by the Auth
```
Similarly, you might say "session recordings backend" here.

zmb3 commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,520 @@
+The load balancer forwards traffic from users and services to an available
+Teleport instance. This must not terminate TLS, and must transparently forward
+the TCP traffic it receives. In other words, this must be a Layer 4 load
+balancer, not a Layer 7 (e.g., HTTP) load balancer. 
```
Worth mentioning cross-zone load balancing here?

zmb3 commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,520 @@
+With TLS Routing, the Teleport Proxy Service uses application-layer protocol
+negotiation (ALPN) to handle all communication with users and services via the
+HTTPS port, regardless of protocol. Without TLS Routing, the Proxy Service
+listens on a separate port for each protocol.
```
```suggestion
listens on separate ports for each protocol.
```

zmb3 commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,520 @@
+events:
+
+- Amazon DynamoDB
+- Google Cloud Firestore
```
I see etcd mentioned below, but why is it not included in this list?

ptgott replied:
I removed it since this is a list of solutions we support for cluster state and audit events, and we don't support etcd for audit events (https://github.com/gravitational/teleport/pull/19217#discussion_r1044817411). It does make sense to put etcd in this list, though, so I'll add it with a parenthetical to indicate we don't support it for audit events.

zmb3 commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,520 @@
+<Admonition title="Required permissions">
+
+In your cloud provider's RBAC solution, your Auth Service instances need
+permissions to create, read from, and write to buckets.
```
For GCS we also need delete, since we 'emulate' S3's multipart API :-(

zmb3 commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,520 @@
+
+## TLS credential provisioning 
+
+High-availability Teleport deployments require an automated system to fetch TLS
```
Technically we don't require an _automated_ system. You can manually procure TLS certificates from any CA you'd like and still deploy Teleport in a highly available manner.

LetsEncrypt or ACM are certainly more convenient, but I'm not sure they impact the availability properties of the system (so long as you don't let your certs expire)

hugoShaka replied:
Automatic cert-provisioning impacts the availability properties negatively because we poorly handle certificate reload. It requires breaking all existing sessions, + the Helm chart does not automatically reload certs, and Teleport never re-reads its files. Running with a good ol' DigiCert wildcard valid for 3 years provides better availability than using LE.

ptgott replied:
👍

zmb3 commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,520 @@
+DynamoDB, Google Cloud Firestore, and etcd as the Auth Service backend:
+
+<Tabs>
+<TabItem label="Amazon DynamoDB">
```
A lot of this is covered in more detail in the backends doc page. Should we be trying to maintain a single source of truth instead of duplicating that info here?

zmb3 commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,520 @@
+    project_id: my-google-cloud-project
+    collection_name: my-cluster-state-table
+    credentials_path: /etc/teleport-secrets/gcp-credentials.json
+    audit_events_uri: ['firestore://my-audit-events-table?projectID=my-google-cloud-project&credentialsPath=/etc/teleport-secrets/gcp-credentials.json', 'stdout://']
```
It's very important that this is a different collection than the one mentioned in `collection_name`. We cover this in the backends doc page, so I'm concerned that duplicating the information here (in less detail) will cause readers to never see the other page and miss important details.

zmb3 commented on docs/pages/deploy-a-cluster/high-availability.mdx:
````diff
@@ -0,0 +1,520 @@
+Teleport configuration:
+
+```yaml
+version: v2
````
`v3` is the latest config version, and to my knowledge the docs were all updated to use v3 where possible. Can we stick with v3 here as well?

zmb3 commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -0,0 +1,520 @@
+teleport:
+  storage:
+  # ...
+  auth_service:
```
These sections are at the root, not under `teleport`.

ptgott commented:
TODO (in addition to the above):
- Include the `docs/pages/includes/cloud/call-to-action.mdx` partial at the top of the guide
- Don't forget the `forScopes` config

ptgott committed:
Respond to zmb3 feedback

- Use "cluster state backend" and "session recording backend"
- Mention cross-zone load balancing
- Link to the Backends Reference instead of including example backend
  configurations
- Use v3 for example configs
- Fix example config indentation

ptgott committed:
Add CTA for Teleport Cloud and forScopes
